### PR TITLE
Add OpenSearch

### PR DESCRIPTION
**Pull request by nicoloboschi**

### CODE REVIEW

**Review by eolivelli — APPROVED** [review]
LGTM

please search all the references to Milvus and add the links to OpenSearch next to them, there are a few more places

you must also link this page from the main summary

**Comment by nicoloboschi** [comment]
Done thanks @eolivelli

File: configuration-resources/data-storage/opensearch.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
+# OpenSearch
+
+LangStream allows you to use OpenSearch as a vector database.
```
Comment (mendonk) [inline configuration-resources/data-storage/opensearch.md:3]:
```suggestion
LangStream supports using OpenSearch as a vector database. 
```

File: configuration-resources/data-storage/opensearch.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
+# OpenSearch
+
+LangStream allows you to use OpenSearch as a vector database. 
+
+You can find more about how to perform Vector Search with OpenSearch in the [official documentation](https://opensearch.org/docs/latest/search-plugins/knn/index/)
```
Comment (mendonk) [inline configuration-resources/data-storage/opensearch.md:5]:
```suggestion
Learn more about performing vector search with OpenSearch in the [official documentation](https://opensearch.org/docs/latest/search-plugins/knn/index/)
```

File: configuration-resources/data-storage/opensearch.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+      configuration:
+        service: "opensearch"
+        username: "${secrets.opensearch.username}"
+        password: "${secrets.opensearch.password}"
+        host: "${secrets.opensearch.host}"
+        region: "${secrets.opensearch.region}"
+        index-name: "my-index-000"
+```
+
+- `username` is the AWS Access Key.
```
Comment (mendonk) [inline configuration-resources/data-storage/opensearch.md:42]:
```suggestion
- `username` is the AWS Access Key
```

File: configuration-resources/data-storage/opensearch.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+- `password` is the AWS Secret Key.
```
Comment (mendonk) [inline configuration-resources/data-storage/opensearch.md:43]:
```suggestion
- `password` is the AWS Secret Key
```

File: configuration-resources/data-storage/opensearch.md [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+- `host` is the endpoint provided by AWS. e.g. for AWS OpenSearch serverless it looks like this: xxxx.<region>.aoss.amazonaws.com
+- `region` is the AWS region. It has to match with the one used in the endpoint.
```
Comment (mendonk) [inline configuration-resources/data-storage/opensearch.md:45]:
```suggestion
- `region` is the AWS region. It has to match with the one used in the endpoint
```

File: configuration-resources/data-storage/opensearch.md [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+
+
+#### Declare an index as asset
+
+To bind the application startup to the OpenSearch index creation, you need to use the `opensearch-index` asset type.
```
Comment (mendonk) [inline configuration-resources/data-storage/opensearch.md:51]:
```suggestion
To bind the application to the OpenSearch index creation at startup, you must use the `opensearch-index` asset type.
```

File: configuration-resources/data-storage/opensearch.md [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        }
+```
+
+Refer to the [settings](https://opensearch.org/docs/latest/im-plugin/index-settings/) documentation for the `settings` field.
+Refer to the [mappings](https://opensearch.org/docs/latest/field-types/index/) documentation for the `mappings` field.
+
+
+#### Search
+
+Use the `query-vector-db` agent to perform searches on the index, with the following parameters:
```
Comment (mendonk) [inline configuration-resources/data-storage/opensearch.md:89]:
```suggestion
Use the `query-vector-db` agent with the following parameters to perform searches on the index created above :
```

File: configuration-resources/data-storage/opensearch.md [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+              }
+            }
+          }
+        }
+      fields:
+        - "value.question_embeddings"
+      output-field: "value.related_documents"
+```
+
+As usual you can use the '?' symbol as a placeholder for the fields.
```
Comment (mendonk) [inline configuration-resources/data-storage/opensearch.md:113]:
```suggestion
You can use the '?' symbol as a placeholder for the fields.
```

File: configuration-resources/data-storage/opensearch.md [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+The `query` is the body sent to OpenSearch. Refer to the [documentation](https://opensearch.org/docs/latest/query-dsl/index/) to learn which parameters are supported.
+Note that the query will be executed on the configured index. Multi-indexes queries are not supported, altough you can declare multiple datasources and deal with different indexes in the same application.
```
Comment (mendonk) [inline configuration-resources/data-storage/opensearch.md:116]:
```suggestion
Note that the query will be executed on the configured index. Multi-index queries are not supported, but you can declare multiple datasources and query different indexes in the same application.
```

File: configuration-resources/data-storage/opensearch.md [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+
+The `output-field` will contain the result.
```
Comment (mendonk) [inline configuration-resources/data-storage/opensearch.md:118]:
```suggestion
The `output-field` will contain the query result. 
```

File: configuration-resources/data-storage/opensearch.md [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+      output-field: "value.related_documents"
+```
+
+As usual you can use the '?' symbol as a placeholder for the fields.
+
+The `query` is the body sent to OpenSearch. Refer to the [documentation](https://opensearch.org/docs/latest/query-dsl/index/) to learn which parameters are supported.
+Note that the query will be executed on the configured index. Multi-indexes queries are not supported, altough you can declare multiple datasources and deal with different indexes in the same application.
+
+The `output-field` will contain the result. 
+The result is in form of array, where each item contains:
```
Comment (mendonk) [inline configuration-resources/data-storage/opensearch.md:119]:
```suggestion
The result is an array with the following elements:
```

File: configuration-resources/data-storage/opensearch.md [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+- `id`: the document ID
+- `document`: the document source 
+- `score`: the document score
+- `index`: the index name
+
+For example, if you want keep only a relevant field from the first result you can use the `compute` agent after the search:
```
Comment (mendonk) [inline configuration-resources/data-storage/opensearch.md:125]:
```suggestion
For example, if you want to keep only one relevant field from the first result, use the `compute` agent after the search:
```

File: configuration-resources/data-storage/opensearch.md [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        - name: "id"
+          expression: "fn:concat(value.filename, value.chunk_id)"
+        - name: "embeddings"
+          expression: "fn:toListOfFloat(value.embeddings_vector)"
+        - name: "text"
+          expression: "value.text"
+```
+
+
+All the indexes are performed using the Bulk operation.
```
Comment (mendonk) [inline configuration-resources/data-storage/opensearch.md:173]:
```suggestion
All indexing is performed using the Bulk operation.
```

**Review by mendonk — APPROVED** [review]
+1

**Pull request merged**